File: s3/s3-client-annotation-processor/src/main/java/ru/tinkoff/kora/s3/client/annotation/processor/S3Operation.java
package ru.tinkoff.kora.s3.client.annotation.processor;

import com.squareup.javapoet.CodeBlock;

import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.ExecutableElement;

public record S3Operation(ExecutableElement method,
                          AnnotationMirror annotation,
                          OperationType type,
                          ImplType impl,
                          Mode mode,
                          CodeBlock code) {

    public enum Mode {
        ASYNC,
        SYNC
    }

    public enum OperationType {
        GET,
        LIST,
        PUT,
        DELETE
    }

    public enum ImplType {
        SIMPLE,
        AWS,
        MINIO
    }
}

File: experimental/s3-client-symbol-processor/src/main/kotlin/ru/tinkoff/kora/s3/client/symbol/processor/S3Operation.kt
package ru.tinkoff.kora.s3.client.symbol.processor

import com.google.devtools.ksp.symbol.KSAnnotation
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.squareup.kotlinpoet.CodeBlock

data class S3Operation(
    val method: KSFunctionDeclaration,
    val annotation: KSAnnotation,
    val code: CodeBlock
) {
    enum class OperationType {
        GET,
        LIST,
        PUT,
        DELETE
    }
}
